core/src/main/java/io/grpc/internal/AbstractClientStream.java

```diff
@@ -119,8 +124,14 @@ protected void inboundHeadersReceived(Metadata.Headers headers) {
     }
     if (headers.containsKey(ChannelImpl.MESSAGE_ENCODING_KEY)) {
       String messageEncoding = headers.get(ChannelImpl.MESSAGE_ENCODING_KEY);
-      setDecompressor(messageEncoding);
+      try {
+        setDecompressor(messageEncoding);
+      } catch (IllegalArgumentException e) {
+        inboundTransportError(Status.INVALID_ARGUMENT.withDescription(e.getMessage()));
```

Comment:
`inboundTransportError()` doesn't actually kill the stream. And yes, that means that the other usage of `inboundTransportError()` in this file is broken. It looks like they should probably call `cancel()` instead now-a-days.